### PR TITLE
Document Users API

### LINKED ISSUE #390: Document Users API
#### GET /users

Retrieves a list of users  based on provided filters. Returns a JSON response with an array of properly serialized users

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_" 
            "http://api.lvh.me:5000/users?filter[id]=1,2,3"
```
#### GET /user

Retrieves the current logged in user from the database. Returns a JSON object containing a properly serialized user.

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_" 
            "http://api.lvh.me:5000/user"
```

**_Example Response_**

```
{
  "data": {
    "id": "3",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T20:19:26.452Z",
      "email": "thesamuelralak@gmail.com",
      "name": null,
      "username": "samuelralak",
      "twitter": null,
      "biography": "editted",
      "website": "new.inital.com",
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```
#### GET /users/email_available

Checks whether an email exists in the database and returns its availability.

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_" 
            "http://api.lvh.me:5000/users/email_available?email=thesamuelralak@gmail.com"
```

**_Response when email is valid and exists_**

```
{
  "available": false,
  "valid": true
}
```

**_Response when email is valid and  doesn't exist_**

```
{
  "available": true,
  "valid": true
}
```

**_Response when email is invalid_**

```
{
  "available": true,
  "valid": false
}
```
#### GET /users/username_available

Checks whether a username is valid and exists in the database and returns its availability.

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_"  
            "http://api.lvh.me:5000/users/username_available?username=samuelralak"
```

**_Response when username is valid and exists_**

```
{
  "available": false,
  "valid": true
}
```

**_Response when username is valid and  doesn't exist_**

```
{
  "available": true,
  "valid": true
}
```

**_Response when username is invalid_**

```
{
  "available": true,
  "valid": false
}
```
#### GET /users/:id

Retrieves a registered user from the database. Responds with a 404 status code and proper error messages if user is not found.

**_Example Request_**

```
curl -X GET -H "Accept: application/json" 
            -H "Content-Type: application/json" 
            -H "Authorization: Bearer _insert-access-token-here_"  
            "http://api.lvh.me:5000/users/2"
```

**_Example response when user exists_**

```
{
  "data": {
    "id": "2",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T19:36:37.180Z",
      "email": null,
      "name": null,
      "username": "codecorpsguy",
      "twitter": null,
      "biography": null,
      "website": null,
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": [
          {
            "id": "1",
            "type": "organizations"
          }
        ]
      },
      "organization_memberships": {
        "data": [
          {
            "id": "1",
            "type": "organization_memberships"
          }
        ]
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

**_Response when user is not found_**

```
{
  "errors": [
    {
      "id": "RECORD_NOT_FOUND",
      "title": "Record not found",
      "detail": "Couldn't find User with 'id'=4",
      "status": 404
    }
  ]
}
```
#### POST /users/

Creates a new user and adds to the database if parametes are valid. It will respond with a 422 status code and proper error messages if a user is invalid.

**_Example Request - with email_**

```
curl -X POST -H "Accept: application/json" 
             -H "Content-Type: application/json" 
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "email": "thesamuelralak@gmail.com",
                        "username": "samuelralak",
                        "password": "samiralak1"
                      }
                    }
                  }' "http://api.lvh.me:5000/users"
```

**_Example Valid Response_**

```
{
  "data": {
    "id": "4",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-05T01:49:33.544Z",
      "email": null,
      "name": null,
      "username": "samilralak",
      "twitter": null,
      "biography": null,
      "website": null,
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

**_Example Invalid Response_**

```
{
  "errors": [
    {
      "id": "VALIDATION_ERROR",
      "source": {
        "pointer": "data/attributes/email"
      },
      "detail": "has already been taken",
      "status": 422
    },
    {
      "id": "VALIDATION_ERROR",
      "source": {
        "pointer": "data/attributes/username"
      },
      "detail": "has already been taken",
      "status": 422
    }
  ]
}
```
#### POST /users/forgot_password

Allows registered users to reset their password. Returns a user JSON object if email exists otherwise it will return a 422 status code with proper validation errors.

**_Example Request_**

```
curl -X POST -H "Accept: application/json" 
             -H "Content-Type: application/json" 
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "email": "thesamuelralak@gmail.com"
                      }
                    }
                  }' "http://api.lvh.me:5000/users/forgot_password"
```

**_Example Valid Response_**

```
{
    "data": {
        "id": "3",
        "type": "users",
        "attributes": {
            "created_at": "2016-07-03T20:19:26.452Z",
            "email": "thesamuelralak@gmail.com",
            "name": null,
            "username": "samuelralak",
            "twitter": null,
            "biography": "editted",
            "website": "new.inital.com",
            "facebook_id": null,
            "facebook_access_token": null,
            "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
            "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
            "state": "signed_up",
            "theme": "light"
        },
        "relationships": {
            "categories": {
                "data": []
            },
            "user_categories": {
                "data": []
            },
            "organizations": {
                "data": []
            },
            "organization_memberships": {
                "data": []
            },
            "roles": {
                "data": []
            },
            "user_roles": {
                "data": []
            },
            "skills": {
                "data": []
            },
            "user_skills": {
                "data": []
            }
        }
    }
}
```

**_Example Invalid Response_**

```
{
  "errors": [
    {
      "id": "VALIDATION_ERROR",
      "source": {
        "pointer": "data/attributes/email"
      },
      "detail": "doesn't exist in the database",
      "status": 422
    }
  ]
}
```
#### POST /users/reset_password

Changes a registered user's password if token is valid otherwise it returns a JSON object containing proper error messages and a 422 status code

**_Example Request_**

```
curl -X POST -H "Accept: application/json" 
             -H "Content-Type: application/json" 
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "confirmation_token": "7d2c59211330e9add35e3fbefbaa8e31f14aaca2",
                        "password": "newpassword"
                      }
                    }
                  }' "http://api.lvh.me:5000/users/reset_password"
```

**_Example Valid Response_**

```
{
  "data": {
    "id": "3",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T20:19:26.452Z",
      "email": "thesamuelralak@gmail.com",
      "name": null,
      "username": "samuelralak",
      "twitter": null,
      "biography": "editted",
      "website": "new.inital.com",
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

**_Example Invalid Response_**

```
{
  "errors": [
    {
      "id": "VALIDATION_ERROR",
      "source": {
        "pointer": "data/attributes/password"
      },
      "detail": "couldn't be reset",
      "status": 422
    }
  ]
}
```
#### PATCH /users/:id

Allows admin to perform edits to a registered user.

**_Example Request_**

```
curl -X PATCH  -H "Accept: application/json" 
             -H "Content-Type: application/json"
             -H "Authorization: Bearer _insert-access-token-here_"  
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "email": "thesamuelralak.com",
                        "biography": "editted"
                      }
                    }
                  }' "http://api.lvh.me:5000/users/3
```

**_Example Valid Response_**

```
{
  "data": {
    "id": "3",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T20:19:26.452Z",
      "email": "thesamuelralak@gmail.com",
      "name": null,
      "username": "samuelralak",
      "twitter": null,
      "biography": "editted",
      "website": "new.inital.com",
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

**_Example Invalid Response - when user is Unauthorized_**

```
{
  "errors": [
    {
      "id": "ACCESS_DENIED",
      "title": "Access denied",
      "detail": "You are not authorized to perform this action on users.",
      "status": 401
    }
  ]
}
```
#### PATCH /users/me

Allows user to perform edits to the current logged in user. Return a JSON object containing a properly serialized user

**_Example Request_**

```
curl -X PATCH  -H "Accept: application/json" 
             -H "Content-Type: application/json"
             -H "Authorization: Bearer _insert-access-token-here_"  
             -d '{
                    "data": {
                      "type": "users",
                      "attributes": {
                        "website": test.com
                        "biography": "editted"
                      }
                    }
                  }' "http://api.lvh.me:5000/users/me"
```

**_Example Valid Response_**

```
{
  "data": {
    "id": "3",
    "type": "users",
    "attributes": {
      "created_at": "2016-07-03T20:19:26.452Z",
      "email": "thesamuelralak@gmail.com",
      "name": null,
      "username": "samuelralak",
      "twitter": null,
      "biography": "editted",
      "website": "test.com",
      "facebook_id": null,
      "facebook_access_token": null,
      "photo_thumb_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_thumb.png",
      "photo_large_url": "https://d3pgew4wbk2vb1.cloudfront.net/icons/user_default_large.png",
      "state": "signed_up",
      "theme": "light"
    },
    "relationships": {
      "categories": {
        "data": []
      },
      "user_categories": {
        "data": []
      },
      "organizations": {
        "data": []
      },
      "organization_memberships": {
        "data": []
      },
      "roles": {
        "data": []
      },
      "user_roles": {
        "data": []
      },
      "skills": {
        "data": []
      },
      "user_skills": {
        "data": []
      }
    }
  }
}
```

### PR DESCRIPTION
**Pull request by joshsmith**
Closes #390 when finished.

This is just a first pass. Could use cleaned up and expanded on.

@christopherstyles is going to run with this.

### CODE REVIEW

File: blueprint/api.apib [hunk 1]
```diff
@@ -91,30 +96,34 @@ Exchange credentials for `access_token`.
...
+The API allows you to create and update users. You can retrieve individual users as well as a filtered list of users. You can also search for available emails and usernames. You can create a password reset token and also use that token to reset the user's password.
+
 ## User [/users/{id}]
 
 + Parameters
 
   + id: 1 (string, required)
 
-  + theme (enum[string], optional)
+### Create a user [POST]
```
Comment (npendery) [inline blueprint/api.apib:107]: Shouldn't this be in its own `User` section as `## User [/users/new]`?

File: blueprint/api.apib [hunk 2]
```diff
@@ -165,48 +181,61 @@ Exchange credentials for `access_token`.
...
++ state: `signed_up` (string)
++ theme: `light` (string, optional)
   The theme is either `light` or `dark`.
++ twitter: `codecorps` (string, optional)
++ username: `codecorps` (string, required)
++ website: `https://codecorps.org` (string, optional)
 
 ## User Relationships Base (object)
 + include User Roles Relationships
 + include User Skills Relationships
```
Comment (thomascchen) [inline blueprint/api.apib:210]: Should we also have relationships for `categories`, `skills`, `organizations`, and `organization_memberships`?

**Comment by joshsmith** [comment]
@christopherstyles are you working on this one still?

**Comment by christopherstyles** [comment]
@JoshSmith yyyep, still working on it.

**Comment by joshsmith** [comment]
@christopherstyles should we close this out due to #464?

**Comment by christopherstyles** [comment]
@JoshSmith I think so, feel free to close this one up ;)

**Pull request closed without merging**